new: introducing thread capabilities monitoring

loresuso opened the pull request:
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

/area driver-kmod

/area driver-ebpf

/area libscap

/area libsinsp

> /area tests

> /area proposals


**What this PR does / why we need it**:

This PR aims at introducing support for thread capabilities monitoring. It introduces three new fields, namely 👇 

- `thread.cap_permitted`
- `thread.cap_inheritable`
- `thread.cap_effective`

since capabilities are a per-thread attribute. 

Capabilities state may be affected by various syscalls. I have retrieved useful information from the [capabilities](https://man7.org/linux/man-pages/man7/capabilities.7.html) and [user_namespace](https://man7.org/linux/man-pages/man7/user_namespaces.7.html) man pages, in order to list all the possible ways capabilities may change. To be precise: 

- capabilities are copied from the parent process on `fork`
- capabilities are modified according to a precise set of rules on `execve`, `execveat`. For this reason, I decided to retrieve them from the drivers to properly modify the state
- a process may modify its capabilities using the `capset` system call
- a process entering a new user namespace acquires the full set of capabilities. This may occur when using `clone`, `clone3`, `setns` and `unshare` syscalls.

Monitoring thread capabilities can be useful for many different purposes. For instance, we will be able to monitor which capabilities ara granted to a container at start time, and we will be able to write Falco rules to alert if a powerful capability was granted since some of them were used in the past to perform various kind of attacks and container escapes. 

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` please post the related issues/tests in a comment and do not use `Fixes`.
-->

Helps with https://github.com/falcosecurity/falco/issues/1812

**Special notes for your reviewer**:
When extracting capabilities, I have decided to not introduce libraries such as libcap. We only need to perform a translation from the numeric representation of a capability to a string, so by introducing that library we will end up not using most of its functionalities. 
Also, for the `capset` syscall I decided to retrieve capabilities directly from the kernel to avoid to `copy_from_user`. 

**Does this PR introduce a user-facing change?**:

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of libscap`.
-->

```release-note
new: introduce thread capabilities fields 
```

poiana commented:
Hi @loresuso. Thanks for your PR.

I'm waiting for a [falcosecurity](https://github.com/orgs/falcosecurity/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Flibs).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

FedeDP commented on driver/syscall_table.c:
```diff
@@ -373,6 +373,9 @@ const struct syscall_evt_pair g_syscall_table[SYSCALL_TABLE_SIZE] = {
 #ifdef __NR_copy_file_range
 	[__NR_copy_file_range - SYSCALL_TABLE_ID0] =            {UF_USED, PPME_SYSCALL_COPY_FILE_RANGE_E, PPME_SYSCALL_COPY_FILE_RANGE_X},
 #endif
+#ifdef __NR_capset
```
Is there no `__NR_ia32_capset` or did you forget about it? :D

loresuso replied:
The second one, of course 😄

FedeDP commented:
Hi @loresuso ! Thanks for this great PR!
/ok-to-test

FedeDP commented on userspace/libscap/scap_procs.c:
```diff
@@ -242,13 +284,13 @@ int32_t scap_proc_fill_info_from_stats(scap_t *handle, char* procdirname, struct
 	}
 
-	ASSERT(nfound == 10 || nfound == 7 || nfound == 6);
+	ASSERT(nfound == 13 || nfound == 7 || nfound == 6);
```
I think `nfound == 7 || nfound == 6` is a really weird check; did you understand by chance why is that check there?
Perhaps we could use a bitmask flag here, something like
```
nfound |= TINFO_UID_F;
nfound |= TINFO_CAPEFF_F;

...

if (nfound == TINFO_MAX_F || nfound == TINFO_xx_F || ...)
```

(given that i do not know that `7` and `6` means in the above check! :D

loresuso replied:
Yes, it is indeed a really weird check, but I assumed that only touching the `nfound == 10` would not break the other two checks. I suppose that in old kernels, not all the fields that are nowadays in `/proc/<pid>/status` were there and fields got into that file incrementally, that's why I would stick with that check!

FedeDP replied:
That makes sense; still, i'd like to have some sort of documentation on top of that (just a comment is fine).  
I mean, my proposed code would be self-documenting but i am ok with leaving it as it is!

loresuso commented:
> Hi @loresuso ! Thanks for this great PR! /ok-to-test

Thank you @FedeDP 🙂

mstemm commented on driver/event_table.c:
```diff
@@ -305,7 +305,7 @@ const struct ppm_event_info g_event_info[PPM_EVENT_MAX] = {
 	/* PPME_PAGE_FAULT_X */ {"NA5", EC_OTHER, EF_UNUSED, 0},
 	/* PPME_SYSCALL_EXECVE_19_E */{"execve", EC_PROCESS, EF_MODIFIES_STATE, 1, {{"filename", PT_FSPATH, PF_NA} } },
-	/* PPME_SYSCALL_EXECVE_19_X */{"execve", EC_PROCESS, EF_MODIFIES_STATE, 20, {{"res", PT_ERRNO, PF_DEC}, {"exe", PT_CHARBUF, PF_NA}, {"args", PT_BYTEBUF, PF_NA}, {"tid", PT_PID, PF_DEC}, {"pid", PT_PID, PF_DEC}, {"ptid", PT_PID, PF_DEC}, {"cwd", PT_CHARBUF, PF_NA}, {"fdlimit", PT_UINT64, PF_DEC}, {"pgft_maj", PT_UINT64, PF_DEC}, {"pgft_min", PT_UINT64, PF_DEC}, {"vm_size", PT_UINT32, PF_DEC}, {"v ...
+	/* PPME_SYSCALL_EXECVE_19_X */{"execve", EC_PROCESS, EF_MODIFIES_STATE, 23, {{"res", PT_ERRNO, PF_DEC}, {"exe", PT_CHARBUF, PF_NA}, {"args", PT_BYTEBUF, PF_NA}, {"tid", PT_PID, PF_DEC}, {"pid", PT_PID, PF_DEC}, {"ptid", PT_PID, PF_DEC}, {"cwd", PT_CHARBUF, PF_NA}, {"fdlimit", PT_UINT64, PF_DEC}, {"pgft_maj", PT_UINT64, PF_DEC}, {"pgft_min", PT_UINT64, PF_DEC}, {"vm_size", PT_UINT32, PF_DEC}, {"v ...
```
When adding new fields to an event, I believe we always want to append to the end. This ensures that when reading events from older libs versions, the subset of fields is consistent.

mstemm replied:
thanks for changing it!

mstemm commented on driver/event_table.c:
```diff
@@ -343,11 +343,13 @@ const struct ppm_event_info g_event_info[PPM_EVENT_MAX] = {
 	/* PPME_SYSCALL_MPROTECT_X */{"mprotect", EC_MEMORY, EF_DROP_SIMPLE_CONS, 1, {{"res", PT_ERRNO, PF_DEC} } },	
 	/* PPME_SYSCALL_EXECVEAT_E */{"execveat", EC_PROCESS, EF_MODIFIES_STATE, 3, {{"dirfd", PT_FD, PF_DEC}, {"pathname", PT_FSRELPATH, PF_NA, DIRFD_PARAM(0)}, {"flags", PT_FLAGS32, PF_HEX, execveat_flags} } },
-	/* PPME_SYSCALL_EXECVEAT_X */{"execveat", EC_PROCESS, EF_MODIFIES_STATE, 20, {{"res", PT_ERRNO, PF_DEC}, {"exe", PT_CHARBUF, PF_NA}, {"args", PT_BYTEBUF, PF_NA}, {"tid", PT_PID, PF_DEC}, {"pid", PT_PID, PF_DEC}, {"ptid", PT_PID, PF_DEC}, {"cwd", PT_CHARBUF, PF_NA}, {"fdlimit", PT_UINT64, PF_DEC}, {"pgft_maj", PT_UINT64, PF_DEC}, {"pgft_min", PT_UINT64, PF_DEC}, {"vm_size", PT_UINT32, PF_DEC}, {" ...
+	/* PPME_SYSCALL_EXECVEAT_X */{"execveat", EC_PROCESS, EF_MODIFIES_STATE, 23, {{"res", PT_ERRNO, PF_DEC}, {"exe", PT_CHARBUF, PF_NA}, {"args", PT_BYTEBUF, PF_NA}, {"tid", PT_PID, PF_DEC}, {"pid", PT_PID, PF_DEC}, {"ptid", PT_PID, PF_DEC}, {"cwd", PT_CHARBUF, PF_NA}, {"fdlimit", PT_UINT64, PF_DEC}, {"pgft_maj", PT_UINT64, PF_DEC}, {"pgft_min", PT_UINT64, PF_DEC}, {"vm_size", PT_UINT32, PF_DEC}, {" ...
```
Same here about appending to end.

mstemm commented on userspace/libscap/scap_savefile.c:
```diff
@@ -410,6 +410,9 @@ int32_t scap_write_proclist_entry_bufs(scap_t *handle, scap_dumper_t *d, struct
 		    scap_dump_write(d, &rootlen, sizeof(uint16_t)) != sizeof(uint16_t) ||
                     scap_dump_write(d, (char *) root, rootlen) != rootlen ||
             scap_dump_write(d, &(tinfo->loginuid), sizeof(uint32_t)) != sizeof(uint32_t) ||
+			scap_dump_write(d, &(tinfo->cap_inheritable), sizeof(uint64_t)) != sizeof(uint64_t) ||
```
Same here about order of fields for the event.

mstemm replied:
thanks for changing it!

mstemm commented on userspace/libsinsp/utils.cpp:
```diff
@@ -1537,6 +1537,76 @@ bool sinsp_utils::startswith(const std::string& s, const std::string& prefix)
 }
 
+const vector<string> capabilities {
+	{"CAP_CHOWN"},
```
To retain portability, we generally define ppm-specific values whenever interpreting binary values as strings, even when the binary values have well known mappings. For example, https://github.com/falcosecurity/libs/blob/master/driver/ppm_events_public.h#L43 and https://github.com/falcosecurity/libs/blob/master/driver/ppm_flag_helpers.h#L300.

I think you want to do the same thing here.

mstemm replied:
Thanks!

mstemm reviewed (changes requested):
Thank you! This looks like a really good feature to add.

Added some feedback.

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/234#" title="Author self-approved">loresuso</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **mstemm** after the PR has been reviewed.

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["mstemm"]} -->

loresuso commented:
> Thank you! This looks like a really good feature to add.
> 
> Added some feedback.

Hey @mstemm, thanks for your feedback, really appreciated! I have addressed all your comments, let me know what do you think now 🙂

loresuso commented:
/check-dco

loresuso commented:
/retest

leogr commented:
Hey @loresuso 

Could you rebase, please?

I'd love to have this feature merged. Thank you!

loresuso commented:
Thank you @leogr! Just rebased! 
/retest

FedeDP commented on driver/ppm_events_public.h:
```diff
@@ -998,7 +1043,9 @@ enum ppm_event_type {
 	PPME_SYSCALL_OPEN_BY_HANDLE_AT_E = 336,
 	PPME_SYSCALL_OPEN_BY_HANDLE_AT_X = 337,
-	PPM_EVENT_MAX = 338
+	PPME_SYSCALL_CAPSET_E = 339,
```
338 right? :D

loresuso replied:
Yeah, of course, I am fixing it! By the way, I really don't like having this assignment in a enum. I mean, these constants start from zero and they get increased by one automatically, we don't need to specify the exact value for each one. This may lead to errors like this. I think I am going to propose a PR to fix this if you agree! 😅

FedeDP replied:
You know, that bothered me too: https://github.com/falcosecurity/libs/pull/86 :D

loresuso replied:
Cool!! Nice to see that!
BTW just fixed this 😄  thanks a lot

FedeDP reviewed (approved):
It LGTM! Thanks, wonderful feature!

/approve

loresuso commented:
/test build-libs-minimal

leogr reviewed (dismissed):
/approve

poiana commented:
LGTM label has been added.  <details>Git tree hash: e9cfae5a5ff879afbbe5b4cc30ae21476ca16004</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/234#pullrequestreview-914118763" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/libs/pull/234#pullrequestreview-915534260" title="Approved">leogr</a>*, *<a href="https://github.com/falcosecurity/libs/pull/234#" title="Author self-approved">loresuso</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [FedeDP,leogr]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

FedeDP commented:
@mstemm are you ok with the changes rn?
Can you approve? :D Thanks!

leogr commented:
Hey @mstemm 

Could you PTAL again? This PR is currently blocked by your change request (that should have been addressed).

loresuso commented:
/retest

loresuso commented:
/check-dco

FedeDP reviewed (approved):
/approve

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/234#pullrequestreview-919921295" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/libs/pull/234#pullrequestreview-915534260" title="Approved">leogr</a>*, *<a href="https://github.com/falcosecurity/libs/pull/234#" title="Author self-approved">loresuso</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [FedeDP,leogr]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

FedeDP commented:
Great job @loresuso ! :partying_face:
